maestro-ios-xctest-runner/maestro-driver-iosUITests/Routes/Handlers/ViewHierarchyHandler.swift

```diff
@@ -57,105 +54,110 @@ struct ViewHierarchyHandler: HTTPHandler {
+
+                // When the application element is skipped, try to fetch
+                // the keyboard and alert hierarchies separately.
+                if let element = element as? XCUIApplication {
+                    let keyboard = try? logger.measure(message: "Fetch keyboard hierarchy") {
+                         try keyboardHierarchy(element)
+                    }
+
+                    let alerts = try? logger.measure(message: "Fetch alert hierarchy", {
+                        try fullScreenAlertHierarchy(element)
+                    })
+
+                    return AXElement(children: [
+                        hierarchy,
+                        keyboard,
+                        alerts
+                    ].compactMap { $0 })
                 }
 
-                let alerts = try? logger.measure(message: "Fetch alert hierarchy", {
-                    try fullScreenAlertHierarchy(element)
-                })
-
-                return AXElement(children: [
-                    hierarchy,
-                    keyboard,
-                    alerts
-                ].compactMap { $0 })
+                return hierarchy
+            } else {
+                logger.error("Snapshot failure, cannot return view hierarchy due to \(message)")
+                return nil
             }
-
-            return hierarchy
+        }
+    }
+    
+    private func getErrorCode(message: String) -> String {
+        if message.contains("Error kAXErrorIllegalArgument getting snapshot for element") {
+            return "illegal-argument-snapshot-failure"
+        } else {
+            return "unknown-snapshot-failure"
         }
     }
 
-    func keyboardHierarchy(_ element: XCUIApplication) throws -> AXElement {
+    private func keyboardHierarchy(_ element: XCUIApplication) throws -> AXElement? {
+        if !element.keyboards.firstMatch.exists {
+            return nil
+        }
+        
         let keyboard = try objcTry {
              element.keyboards.firstMatch
         }
 
         return try elementHierarchy(xcuiElement: keyboard)
     }
 
-    func fullScreenAlertHierarchy(_ element: XCUIApplication) throws -> AXElement {
-        let alerts = try objcTry {
-            element.alerts.allElementsBoundByIndex
+    func fullScreenAlertHierarchy(_ element: XCUIApplication) throws -> AXElement? {
+        if !element.alerts.firstMatch.exists {
+            return nil
         }
-
-        let alertHierarchies = try alerts.map { alert in
-            try elementHierarchy(xcuiElement: alert)
+        
+        let alert = try objcTry {
+            element.alerts.firstMatch
         }
 
-        return AXElement(children: alertHierarchies)
+        return try elementHierarchy(xcuiElement: alert)
     }
 
     let useFirstParentWithMultipleChildren = false
-    func findRecoveryElement(_ element: XCUIElement) -> XCUIElement {
+    private func findRecoveryElement(_ element: XCUIElement) -> XCUIElement {
```

Comment:
Right now we just find the next Any element to find the recoveryElement. Should we make this skipping condition stricter? For example: Checking frame size to see we are not skipping some important XCUIElement

Reply:
The only risk I see in finding `recoveryElement` is that with:

```
element
    .children(matching: .any)
    .firstMatch
```
we might risk skipping important elements. But that's a tradeoff also I think it's just not going to be possible and for React native apps that's mostly the view containers for increasing padding

Reply:
If the view frame is checked, then what do we do when the frame size is unexpected? And why would that be better than continuing the algorithm as normal?
Why not check other things, like the accessibilityLabel?
I still don't understand what is special about the frame size that makes it something that should be checked in the view hierarchy recovery algorithm.

Reply:
> If the view frame is checked, then what do we do when the frame size is unexpected? And why would that be better than continuing the algorithm as normal?

If we check the frame size and it's equal to screen size we know that we can safely skip these elements. If let's say it goes unexpected and we need to still skip it for successful operation we do it intentionally. Its better because we would know where our algorithm could go wrong.